### PR TITLE
Add "ember-test-selectors: The road to 1.0" blog post

### PR DESCRIPTION
**Pull request by Turbo87**
aka. a more detailed changelog :P

### CODE REVIEW

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+which version of `ember-cli-babel` you use. Based on that information we inject
+different Babel transforms into the build pipeline so that we can support both
+major Babel versions with the same addon.
+
+
+#### v0.3.4: Support for test-selectors in nested addons
+
+In April we were notified by [Luke Melia](https://github.com/lukemelia) that
+test-selectors in templates in an addon that he extracted were stripped
+unconditionally. It turned out that if ember-test-selectors was used as a
```
Comment (pangratz) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: ember-test-selectors in backticks?

**Review by pangratz — APPROVED** [review]
🔝 👏

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+A few days later Luke provided us with a repository that reproduced his
+problem. Using his reproduction we have been able to fix the issue and it is
+now possible to also use `ember-test-selectors` as a nested addon dependency
+and rely on correct test-selector stripping depending on the build environment.
+
+
+#### v0.3.7: Deprecation of the `testSelector` helper function
+
+Triggered by [Kelly Selden](https://github.com/kellyselden) a
```
Comment (geekygrappler) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: "Kelly Selden triggered a conversation about" Sounds a bit more natural

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+It seemed that the main issue with Ember 1.11/1.12 support was that our
+template AST transforms were failing and as a first attempt we simply disabled
+them completely which resolved the problem. Unfortunately that meant that
+test-selectors were no longer stripped at all from the templates which did not
+seem like a good solution to us so we dug deeper.
+
+Thanks to [ember-try](https://github.com/ember-cli/ember-try) it was very fast
+and easy to try out our addon on many different Ember versions and we quickly
+discovered that the Handlebars AST had changed between Ember 2.12 and 2.13 in
```
Comment (geekygrappler) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: Should this be 2.12 and 2.13 or 1.12 and 1.13?

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+```
+
+but in Ember 2.12 it looks more like this:
+
+```kotlin
+MustacheStatement {
+  path: PathExpression {
+    original: "user-list"
+  }
+  sexpr: SexprStatement {
```
Comment (geekygrappler) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: Sexpr short for sexpression?

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+```
+
+As you can see the AST looks almost similar, but not exactly the same.
+In the end we figured out we could check if the `MustacheStatement` has
+a `sexpr` property, and in that case use the `hash` property inside of it
+instead.
+
+Once we had that conditional in place all our tests were 🍏 again and we were
+able to adjust our range of supported Ember versions all the way down to
+Ember 2.11.
```
Comment (geekygrappler) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: I think throughout you mean 1.x and are saying 2.x by accident? Not really sure though.

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        value: BooleanLiteral {
+          value: true
+        }
+      }
+    ]
+  }
+}
+```
+
+but in Ember 2.12 it looks more like this:
```
Comment (geekygrappler) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: 1.12?

**Review by geekygrappler — APPROVED** [review]

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md (hunk 3 above)
Comment (pangratz) [reply blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: 👍 Good catch!

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md (hunk 4 above)
Comment (Turbo87) [reply blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: I assume it is https://en.wikipedia.org/wiki/S-expression

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md (hunk 5 above)
Comment (Turbo87) [reply blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: 🤦‍♂️

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md (hunk 6 above)
Comment (Turbo87) [reply blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: wtf... I did mistake that three times in a row?! 😂

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md (hunk 4 above)
Comment (geekygrappler) [reply blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: Sad times.

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md (hunk 6 above)
Comment (pangratz) [reply blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: at least there is consistency 😄

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+property on the `user-list` component it will take the value of the
+`data-test-user-list` property on the parent component and pass that as a
+positional parameter to the `user-list` component.
+
+In v0.2.0 of ember-test-selectors we introduced another Handlebars AST
+transform which transforms all such valueless `data-test-foo` instances to
+`data-test-foo=true` by default.
+
+While that seems to work quite well it creates issues in other places as you
+can see in the following template:
```
Comment (marcoow) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: I'd rephrase the _"in other places"_ which seems a bit strange here. Maybe _"While that seems like a convenient addition, it has drawbacks and does not always work as expected, e.g.:"_

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+{% raw %}{{user-list users=users data-test-user-list=true}}{% endraw %}
+```
+
+We will deprecate the usage of valueless `data-test-*` attributes on components
+in an upcoming release and remove the transform completely for the v1.0.0 release.
+
+
+#### v0.3.0: Support for Babel 6
+
+This year (2017) the Babel ecosystem finally moved from Babel 5 to Babel 6,
```
Comment (marcoow) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: I think this should be _"the Ember ecosystem"_.

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+which happened mostly without issues for app developers as it was just
+a dependency upgrade from `ember-cli-babel@5` to `ember-cli-babel@6`. Some
+app developers might think they are still only using Babel 5 for their app,
+but it is very likely that some of their addons are already using Babel 6 under
+the hood since in the Ember ecosystem each addon controls its own transpilation.
```
Comment (marcoow) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: _"in the Ember ecosystem"_ reads a bit repetitive here - I'd just remove it maybe.

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+For addons that rely on Babel transforms the upgrade unfortunately was not
+quite as smooth, as the Babel transforms API had changed significantly with
+Babel 6.
+
+As `ember-test-selectors` relies on Babel transforms we needed to figure out
```
Comment (marcoow) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: Maybe add _"…relies on Babel transforms (for removing all `data-test-…` properties from your app code in production builds) we needed…"_

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+a solution to this problem. We wanted to support both Babel 5 and Babel 6 in
+the same addon instead of having to publish two different variants of the addon
+each targeting a different Babel version.
+
+Fortunately at [EmberConf](http://emberconf.com/) in March we found a solution:
+The addon is now checking the dependencies of your project and figuring out
```
Comment (marcoow) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: _"…now checks … and figures out…"_

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+
+```js
+let foo = '[data-test-foo]';
+let bar = '[data-test-bar="baz"]';
+```
+
+After discussing back and forth and coming up with
+[alternative APIs](https://github.com/simplabs/ember-test-selectors/pull/122)
+we decided the best way forward was actually to not use any helpers at all.
+This has the advantage of not hiding the actual CSS selector that is being used
```
Comment (marcoow) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: I'd say _"…not obfuscating the actual…"_

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+        }
+      ]
+    }
+  }
+}
+```
+
+As you can see the AST looks almost similar, but not exactly the same.
+In the end we figured out we could check if the `MustacheStatement` has
+a `sexpr` property, and in that case use the `hash` property inside of it
```
Comment (marcoow) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: I think _"inside of that"_ is better.

**Review by marcoow — COMMENTED** [review]
Yay, awesome post! Only left a few comments for rephrasing some parts to make the intention a bit clearer.

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+github-handle: Turbo87
+twitter-handle: tobiasbieniek
+---
+
+Back in January we wrote about the
+[latest changes](https://simplabs.com/blog/2017/01/13/ember-test-selectors.html)
+in [`ember-test-selectors`](https://github.com/simplabs/ember-test-selectors)
+and how we implemented them. Since then we adjusted a few things and this blog
+post should give you an idea what has happened so far and what else will happen
+before we feel comfortable promoting the addon to v1.0.0.
```
Comment (jayjayjpg) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: I like that mention of caring about proper semver a lot 👍

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+
+<!--break-->
+
+#### v0.2.0: `data-test-*` attributes without values
```
Comment (jayjayjpg) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: Might it be good if this reads as 
```
Deprecation of data-test-* attributes without values
```
so it already points to the conclusion of this section?

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+#### Conclusion
+
+A lot of things have happened on the project this year and we will continue to
+iterate forward on this. One exciting enhancement that we are looking into is
+adding support for [glimmer.js](https://glimmerjs.com/) to the addon. This will
+likely not be done by the time we release v1.0.0, but will be something for us
+to work on in the next year.
+
+If you have questions on how to best take advantage of test-selectors or how to
+structure your tests in general make sure to [contact us](https://simplabs.com/contact/).
```
Comment (jayjayjpg) [inline blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: :heart:

**Review by jayjayjpg — COMMENTED** [review]
That's an awesome post and I enjoy the detailed overview on which decisions went into improving the addon a lot. Just a few minor comments from my side 👍

File: blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md (hunk 15 above)
Comment (Turbo87) [reply blog/_posts/2017-11-16-ember-test-selectors-road-to-1-0.md]: well the problem is that in `v0.2.0` we implemented them and they are not actually deprecated yet, the blog post only mentions that a future release will deprecate them

**Comment by Turbo87** [comment]
Feedback addressed, thanks!

File: blog/_posts/2017-11-17-ember-test-selectors-road-to-1-0.md [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+
+```js
+let foo = '[data-test-foo]';
+let bar = '[data-test-bar="baz"]';
+```
+
+After discussing back and forth and coming up with
+[alternative APIs](https://github.com/simplabs/ember-test-selectors/pull/122)
+we decided the best way forward was actually to not use any helpers at all.
+This has the advantage of not hiding the actual CSS selector that is being used
```
Comment (marcoow) [inline blog/_posts/2017-11-17-ember-test-selectors-road-to-1-0.md:132]: I still think _"obfuscating"_ is better

Comment (Turbo87) [reply blog/_posts/2017-11-17-ember-test-selectors-road-to-1-0.md:132]: yeah, I don't... 😉

Comment (marcoow) [reply blog/_posts/2017-11-17-ember-test-selectors-road-to-1-0.md:132]: Fine, doesn't matter much - please either address or reply to review comments though.

**Review by marcoow — APPROVED** [review]
